### PR TITLE
ability to hide tabs

### PR DESCRIPTION
**Pull request by iantrich**
This allows you to hide tabs outright or by user

```yaml
views:
  - title: Ian
    visible:
      user: iantrich
    cards:
      ...
  - title: Chelsea
    visible:
      user: cswehla
    cards:
      ...
  - title: Admin
    visible: false
    cards:
      ...
```

Closes https://github.com/home-assistant/ui-schema/issues/144

I didn't include a UI editor change for this; I went through a few designs but none that worked great. Will have to think more on that.

Docs: https://github.com/home-assistant/home-assistant.io/pull/10458

### CODE REVIEW

**Comment by iantrich** [comment]
@bramkragten going to merge tonight unless you have any feedback. Thanks.

File: src/panels/lovelace/hui-root.ts [hunk 1]
```diff
@@ -228,7 +228,18 @@ class HUIRoot extends LitElement {
                   >
                     ${this.lovelace!.config.views.map(
                       (view) => html`
-                        <paper-tab aria-label="${view.title}">
+                        <paper-tab
+                          aria-label="${view.title}"
+                          class="${classMap({
+                            "hide-tab": Boolean(
+                              !this._editMode &&
```
Comment (bramkragten) [inline src/panels/lovelace/hui-root.ts]:
```suggestion
                              !this._editMode &&
                                view.hide && view.hide === true ||
                                (Array.isArray(view.hide) &&
                                    view.hide.includes(this.hass!.user!.name)))```

**Comment by bramkragten** [comment]
What would be the use case of `hide: true`?

For a future PR:
- We only hide the tab now, you can still access it with the id/url.

**Comment by iantrich** [comment]
@bramkragten `hide: true` would be something I would use for casting, as it doesn't support dialogs, I plan to create views that act like dialogs that I'll navigate to, but don't necessarily want to have as a tab

**Comment by bramkragten** [comment]
OK, so that would be 2 different functionalities?
- Restrict access to users
- Only be able to access the view by url

Should we combine those in 1 setting? Shouldn't we have a hide in navigation boolean, and a setting to restrict a view to users?

**Comment by iantrich** [comment]
This functionality is only for the viewing a tab. A `restrictions` config sounds like a great idea and would override any `hide` settings

**Comment by bramkragten** [comment]
If it is just for hiding it in the menu, I think it should just be a boolean and not an array of usernames. That should be handled by the `restrictions` config.

**Comment by iantrich** [comment]
No, it should be allowed by username as well. e.g. I want to build out navigation for a casted view to other views since more-info dialogs don't work on cast (how I currently use CCH)
Or I have a view of my automations to turn on/off that I don't want to clutter up my top bar, but still be accessible. It makes sense to support it by user as well.

**Comment by iantrich** [comment]
@bramkragten any further concerns for this? Was hoping to get it in 0.100; for my own selfish reasons as I can then remove CCH 😄

**Comment by bramkragten** [comment]
@balloob what do you think? Boolean or also user based?

**Comment by iantrich** [comment]
Maybe I'm not explaining this correctly, but being user-based is a big part of this.
Here's a simple example maybe assuming I use different users for desktop vs phone:
On my desktop I would want my automations view to be shown all the time in the bar
On my phone I want my automations view to be hidden, but still accessible

**Comment by balloob** [comment]
If user based, it should be user_id based. Username is only for users using the default homeassistant provider, but there are others too and usernames can change. User Id is the only thing that we can trust to remain static.

I also don't know if this is the right approach. With this approach we're going to embed user permissions within Lovelace config. An alternative approach would be to have different Lovelace configs, one per user.

File: src/panels/lovelace/hui-root.ts [hunk 2]
```diff
@@ -228,7 +228,18 @@ class HUIRoot extends LitElement {
...
+                                view.hide &&
+                                (view.hide === true ||
+                                  (Array.isArray(view.hide) &&
+                                    view.hide.includes(this.hass!.user!.name)))
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts]: This value can be changed in the UI. Should be user id.

File: src/panels/lovelace/hui-root.ts [hunk 3]
```diff
@@ -228,7 +228,18 @@ class HUIRoot extends LitElement {
...
                       (view) => html`
-                        <paper-tab aria-label="${view.title}">
+                        <paper-tab
+                          aria-label="${view.title}"
+                          class="${classMap({
+                            "hide-tab": Boolean(
+                              !this._editMode &&
+                                view.hide &&
+                                (view.hide === true ||
+                                  (Array.isArray(view.hide) &&
```
Comment (balloob) [inline src/panels/lovelace/hui-root.ts]: We should be able to assume that if `hide` is truthy, that it's a boolean or an array of strings, so no need to check if it's an array ?

File: src/panels/lovelace/hui-root.ts (hunk 2 above)
Comment (iantrich) [reply src/panels/lovelace/hui-root.ts]: How do I grab that?

**Comment by iantrich** [comment]
A per user lovelace config opens a lot of complications. On mobile here, so quickly something that comes to mind is wanting to restrict devices to certain users. E.g. I don't my roommate to have access to my bedroom light. I think this approach is safe for now and satisfied a lot of what people want.

**Comment by balloob** [comment]
But then what about different tabs for different devices ? etc. 

Once we add a feature, we can't take it back. We need to make sure we think this through and don't jump the gun. We need to support this forever.

**Comment by iantrich** [comment]
> We need to support this forever.

Fair point but both would be hindered by device types and the only way to currently overcome that with either would be separate users. We could add some mobile vs desktop detection or parsing of user agents perhaps; user agents would be most flexible but perhaps not as easy to convey to a user :man_shrugging: I think that could be added to the admin-based hide method fairly easy in terms of a configuration; not entirely sure on execution though, but the same is true for user-based.

```yaml
views:
  - title: Admin
    hide: 
      - user: iantrich
        device: iphone
    cards:
      ...
```

Off the top of my head;

Cons for admin-based
1. Individual users cannot customize the look and feel of their UI. The admin would have to do it for them.

Cons for user-based
1. Individual users could have access to devices an admin might not want them to have access to (a method to restrict devices could be devised on the backend to solve this)
2. Maintenance could be very burdensome. In my case, I would then be responsible for making changes to numerous interfaces for a single change I want to make. Some sort of sharing of views or cards could be employed, but what does that look like/work? Sounds hard to maintain as well. e.g. I make a card and want to share it with everyone, I add some flag or list of userids to my card? Make it public to use? If I share something, does it automatically show up on another users view? How do they reference it? What if they want to modify what I shared? Is that reflected on my view as well?

Perhaps there is some middle ground between the two that I'm not seeing, but I personally don't see a user-based system being a functional option without some serious compromises made.

**Comment by balloob** [comment]
I think first and foremost, we should probably reverse it. Have a show option, so that adding a new user doesn't mean updating all your views.

I do like the idea of making it an object per list item, so that you can specify the type. Allows for easy future extension.

So I guess we get something like this?

```yaml
views:
  - title: Admin
    show_for: 
      - user: abc2de1fghk513i4j1klmno1pq
    cards:
      ...
```

**Comment by iantrich** [comment]
Yeah, that logic makes sense. So to get the user id I need to get the credentials, is that right?

https://developers.home-assistant.io/docs/en/auth_index.html

@bramkragten do you happen to know?

**Comment by balloob** [comment]
`hass.user.id`

**Comment by balloob** [comment]
Also have access to `is_admin` there .

**Comment by iantrich** [comment]
that's it?!? 🙈 lol

File: src/panels/lovelace/hui-root.ts (hunk 3 above)
Comment (iantrich) [reply src/panels/lovelace/hui-root.ts]: Someone could mess it up and make an object, I suppose?

File: src/panels/lovelace/hui-root.ts [hunk 4]
```diff
@@ -228,7 +228,21 @@ class HUIRoot extends LitElement {
...
+                          class="${classMap({
+                            "hide-tab": Boolean(
+                              !this._editMode &&
+                                view.show !== undefined &&
+                                ((Array.isArray(view.show) &&
+                                  !view.show.some(
+                                    (e) => e.user === this.hass!.user!.id
+                                  )) ||
+                                  (typeof view.show === "boolean" &&
+                                    view.show === false))
```
Comment (bramkragten) [inline src/panels/lovelace/hui-root.ts]: `view.show === false` will only match booleans, the triple `=` also checks type, the `==` doesn't.

File: src/data/lovelace.ts [hunk 5]
```diff
@@ -18,6 +18,11 @@ export interface LovelaceViewConfig {
   theme?: string;
   panel?: boolean;
   background?: string;
+  show?: boolean | ShowViewConfig[];
```
Comment (balloob) [inline src/data/lovelace.ts]: I don't like `show` . Show what? I think that it's more descriptive if we make it `show_if`

Comment (balloob) [reply src/data/lovelace.ts]: Wonder if we should change it from `show` to `visible`.

`visible: false` and

```yaml
visible:
 - user: abcd
```

both clearly describe intention of the config option.

Comment (iantrich) [reply src/data/lovelace.ts]: Agreed, `visible` is much better/clearer

**Review by balloob — APPROVED** [review]
Love the simple implementation. People will love this feature, great job Ian! <3

**Pull request merged**